### PR TITLE
Module settings models

### PR DESCRIPTION
**Pull request by allenwq**
Support turn on/off modules in instances/courses.
Add a ModuleSettingsConcern in this PR, which lets instance and course have following functions:
```ruby
@instance.modules # all the available modules 
@instance.enabled_modules
@instance.disabled_modules

# If a module was disabled in that instance, course will not able to see it
@course.modules
@course.enabled_modules
@course.disabled_modules
```

### CODE REVIEW

File: app/models/concerns/module_preferences_concern.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+module ModulePreferencesConcern
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :module_preference_objects, as: :target, dependent: :destroy
```
Comment (lowjoel) [inline app/models/concerns/module_preferences_concern.rb]: `has_many :module_preferences`?

File: app/models/concerns/module_preferences_concern.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    accepts_nested_attributes_for :module_preference_objects
+
+    if self == Course
```
Comment (lowjoel) [inline app/models/concerns/module_preferences_concern.rb]: This is bad. I don't know what this is for though.

File: db/migrate/20150325094527_create_module_preference_objects.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CreateModulePreferenceObjects < ActiveRecord::Migration
+  def change
+    create_table :module_preference_objects do |t|
+      t.references :target, null: false, polymorphic: true
+      t.string :key, null: false
+      t.boolean :enabled, null: false
```
Comment (lowjoel) [inline db/migrate/20150325094527_create_module_preference_objects.rb]: I'm not sure this makes sense...

File: app/models/concerns/module_preferences_concern.rb (hunk 2 above)
Comment (allenwq) [inline app/models/concerns/module_preferences_concern.rb]: Any better way to do this ? I've tried few different methods actually.

Basically, Course and instance need to include this concern. this function is different for them, other functions are same.

**Comment by allenwq** [comment]
@lowjoel I've updated here

File: app/models/concerns/module_settings_concern.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+module ModuleSettingsConcern
```
Comment (lowjoel) [inline app/models/concerns/module_settings_concern.rb]: Would it make sense if this file is in the app/models/concerns/course folder instead?

Comment (lowjoel) [inline app/models/concerns/module_settings_concern.rb]: Oh, because it's used for both course and instance. hmmmmm.

Comment (lowjoel) [inline app/models/concerns/module_settings_concern.rb]: This isn't really a concern about settings per se... is it?

File: lib/autoload/modular.rb [hunk 5]
```diff
@@ -136,5 +136,19 @@ def class_in_path(relative_path)
...
+      # @return [Boolean] the default enabled status of the module
+      def enabled_by_default?
+        true
+      end
+
+      # Unique key of the module, to serve as the key in module_preference_objects and translations
+      #
+      # @return [Symbol] the key
+      def key
+        to_s.underscore.sub('/', '_').to_sym
```
Comment (lowjoel) [inline lib/autoload/modular.rb]: `name.underscore...`, `to_s` assumes that Ruby converts singletons to names.

File: spec/models/course_spec.rb [hunk 6]
```diff
@@ -16,5 +16,94 @@
...
+              course.settings(sample_module.key).enabled = true
+            end
+
+            it 'does not include the enabled module' do
+              expect(disabled_modules.include?(sample_module)).to eq false
+            end
+          end
+        end
+      end
+    end
```
Comment (lowjoel) [inline spec/models/course_spec.rb]: Does this PR handle disabling of course modules if the instance has the module disabled?

File: app/models/course.rb [hunk 7]
```diff
@@ -13,6 +15,11 @@ class Course < ActiveRecord::Base
   has_many :announcements, inverse_of: :course, dependent: :destroy
   has_many :achievements, inverse_of: :course, dependent: :destroy
 
+  # @return [Array] array of all available modules
+  def modules
+    ActsAsTenant.current_tenant.enabled_modules
```
Comment (lowjoel) [inline app/models/course.rb]: Maybe we should extract this out to be an Instance method...? `Instance.current.enabled_modules`

File: app/models/instance.rb [hunk 8]
```diff
@@ -21,4 +23,9 @@ def find_tenant_by_host(host)
   has_many :users, through: :instance_users
 
   has_many :announcements, class_name: Instance::Announcement.name
+
+  # @return [Array] array of all available modules
+  def modules
+    Course::CoursesController.modules
```
Comment (lowjoel) [inline app/models/instance.rb]: this doesn't look right....

File: spec/models/course_spec.rb (hunk 6 above)
Comment (allenwq) [inline spec/models/course_spec.rb]: @lowjoel Updated, have look !

File: app/models/concerns/module_settings_concern.rb (hunk 4 above)
Comment (lowjoel) [inline app/models/concerns/module_settings_concern.rb]: Maybe this should be called EnableableModuleHost or something like this. The fact that a host can store settings should be declared on the module host itself.

Comment (allenwq) [inline app/models/concerns/module_settings_concern.rb]: I guess that one would be another concern ? Cause this concern is for both instance and course, for them to disable/enable modules.

**Comment by lowjoel** [comment]
Replace with #190.

**Pull request closed without merging**